python/ray/air/tests/test_arrow.py

```diff
@@ -10,13 +13,64 @@
     convert_to_pyarrow_array,
 )
 from ray.air.util.tensor_extensions.utils import create_ragged_ndarray
+from ray.tests.conftest import *  # noqa
 
 
 @dataclass
 class UserObj:
     i: int = field()
 
 
+@pytest.mark.parametrize(
+    "numpy_precision, expected_arrow_unit",
+    [
+        # The lowest resolution Arrow supports is seconds.
+        ("Y", "s"),
+        ("M", "s"),
+        ("D", "s"),
+        ("h", "s"),
+        ("m", "s"),
+        ("s", "s"),
+        ("ms", "ms"),
+        ("us", "us"),
+        ("ns", "ns"),
+        # The highest resolution Arrow supports is nanoseconds.
+        ("ps", "ns"),
+        ("fs", "ns"),
+        ("as", "ns"),
+    ],
+)
+def test_convert_datetime_array(
+    numpy_precision: str,
+    expected_arrow_unit: str,
+):
+    numpy_array = np.zeros(1, dtype=f"datetime64[{numpy_precision}]")
+
+    pyarrow_array = _convert_to_pyarrow_native_array(numpy_array, "")
+
+    assert pyarrow_array.type.unit == expected_arrow_unit
+    assert len(numpy_array) == len(pyarrow_array)
+
+
+@pytest.mark.parametrize("dtype", ["int64", "float64", "datetime64[ns]"])
+def test_infer_type_does_not_leak_memory(dtype):
+    # Test for https://github.com/apache/arrow/issues/45493.
+    column_values = np.zeros(923040, dtype=dtype)  # A ~7 MiB column
+
+    process = psutil.Process()
+    gc.collect()
+    pa.default_memory_pool().release_unused()
+    before = process.memory_info().rss
+
+    _infer_pyarrow_type(column_values)
+
+    gc.collect()
+    pa.default_memory_pool().release_unused()
+    after = process.memory_info().rss
+
+    assert after - before < 1024 * 1024, after - before
```

Comment:
Does it fail w/o your changes? What's amount of memory leaked per call?

Reply:
Yeah. Like 130 MiB lol

Reply:
If you haven't already, see the repro in the commented issue: https://github.com/apache/arrow/issues/45493